src/Servers/Kestrel/test/InMemory.FunctionalTests/BadHttpRequestTests.cs

```diff
@@ -140,6 +141,31 @@ public Task BadRequestIfHostHeaderDoesNotMatchRequestTarget(string requestTarget
             CoreStrings.FormatBadRequest_InvalidHostHeader_Detail(host.Trim()));
     }
 
+    [Fact]
+    public async Task CanOptOutOfBadRequestIfHostHeaderDoesNotMatchRequestTarget()
+    {
+        var receivedHost = StringValues.Empty;
+        await using var server = new TestServer(context =>
+        {
+            receivedHost = context.Request.Headers.Host;
+            return Task.CompletedTask;
+        }, new TestServiceContext(LoggerFactory)
+        {
+            ServerOptions = new KestrelServerOptions()
+            {
+                AllowUnsafeHostHeaderOverride = true,
+            }
+        });
+        using var client = server.CreateConnection();
+
+        // Note the missing line-reak between the Host and Connection headers
+        await client.SendAll($"GET http://www.foo.com/api/data HTTP/1.1\r\nHost: www.foo.comConnection: keep-alive\r\n\r\n");
+
+        await client.Receive("HTTP/1.1 200 OK");
+
+        Assert.Equal("www.foo.com:80", receivedHost);
```

Comment:
Just making the difference more pronounced...
```suggestion
        await client.SendAll($"GET http://www.foo.com/api/data HTTP/1.1\r\nHost: www.bar.comConnection: keep-alive\r\n\r\n");

        await client.Receive("HTTP/1.1 200 OK");

        Assert.Equal("www.foo.com:80", receivedHost);
```

Reply:
Should there be a `\r\n` before `Connection:`?

Reply:
Nope - there's a comment to that effect just above.  That's an example of a request that we don't want to reject.

Reply:
@Tratcher I feel like that sort of buries the lede.  Maybe there should be two tests: one with an explicitly different host and one with a missing line-break.

Reply:
Sure, make it a theory with two inputs. The result should be the same.